File: src/emitter/markdown/format-depth-query.ts
import { DEFAULT_DEPTH } from "../../cli/args/depth-options.js";
import type { NestingDepths } from "../../ir/annotations/scope-annotation.js";
import {
  NESTING_KIND,
  type NestingKind,
} from "../../serializer/nesting-kind.js";

const NON_FUNCTION_KINDS: readonly NestingKind[] = [
  NESTING_KIND.If,
  NESTING_KIND.For,
  NESTING_KIND.While,
  NESTING_KIND.Switch,
  NESTING_KIND.TryCatchFinally,
  NESTING_KIND.Block,
];

// Render the CLI invocation that would have produced these per-kind
// thresholds. Returns null when every kind is at DEFAULT_DEPTH (i.e.
// the user did not narrow anything). The CLI surface is just --depth /
// --depth-function / --depth-block, so this picks the shortest
// equivalent of those three; when the non-function kinds disagree
// (programmatic API only, not reachable from the CLI) we fall back to a
// per-kind sh-comment listing so the snapshot still records exactly
// what was applied.
export function formatDepthQuery(
  depths: NestingDepths | undefined,
): string | null {
  if (!depths) {
    return null;
  }
  const fn = depths[NESTING_KIND.Function];
  const blockValues = NON_FUNCTION_KINDS.map((k) => depths[k]);
  const firstBlock = blockValues[0] ?? DEFAULT_DEPTH;
  const blockUniform = blockValues.every((v) => v === firstBlock);

  if (blockUniform) {
    const block = firstBlock;
    if (fn === DEFAULT_DEPTH && block === DEFAULT_DEPTH) {
      return null;
    }
    if (fn === block) {
      return `--depth ${fn}`;
    }
    if (fn === DEFAULT_DEPTH) {
      return `--depth-block ${block}`;
    }
    if (block === DEFAULT_DEPTH) {
      return `--depth-function ${fn}`;
    }
    return `--depth-function ${fn} --depth-block ${block}`;
  }

  // Non-uniform non-function kinds -- only reachable via the
  // programmatic API. Surface every kind so the snapshot is faithful
  // even though no exact CLI flag string can produce it.
  const parts: /* mutable */ string[] = [];
  if (fn !== DEFAULT_DEPTH) {
    parts.push(`function=${fn}`);
  }
  for (const k of NON_FUNCTION_KINDS) {
    if (depths[k] !== DEFAULT_DEPTH) {
      parts.push(`${k}=${depths[k]}`);
    }
  }
  if (parts.length === 0) {
    return null;
  }
  return `# nesting depth: ${parts.join(" ")}`;
}

File: crates/unsnarl-emitter-markdown/src/format_depth_query.rs
//! `format_depth_query`: render the `--depth*` form for the markdown
//! `## Query` section, returning `None` when every kind is still at
//! `DEFAULT_DEPTH` (the user did not narrow anything).
//!
//! The CLI surface exposes only `--depth`, `--depth-function`, and
//! `--depth-block`, so when the seven per-kind values reduce to a
//! `{function, block}` pair this picks the shortest equivalent of
//! those three flags. When the non-function kinds disagree (which is
//! only reachable through the programmatic API, not the CLI) the
//! function delegates to `format_depth_query_per_kind::format`,
//! which lives in its own file so the coverage report can isolate
//! the CLI-unreachable fallback from this CLI-driven entry point.

mod format_depth_query_per_kind;

use unsnarl_emitter::DEFAULT_DEPTH;
use unsnarl_ir::nesting_kind::{NestingDepth, NestingDepths};

pub fn format_depth_query(depths: Option<&NestingDepths>) -> Option<String> {
    let depths = depths?;
    let function = depths.function;
    let block_values: [NestingDepth; 6] = [
        depths.r#if,
        depths.r#for,
        depths.r#while,
        depths.switch,
        depths.try_catch_finally,
        depths.block,
    ];
    let first_block = block_values[0];
    let block_uniform = block_values.iter().all(|v| *v == first_block);

    if !block_uniform {
        return format_depth_query_per_kind::format(depths, function);
    }
    let block = first_block;
    if function == DEFAULT_DEPTH && block == DEFAULT_DEPTH {
        return None;
    }
    if function == block {
        return Some(format!("--depth {}", function.0));
    }
    if function == DEFAULT_DEPTH {
        return Some(format!("--depth-block {}", block.0));
    }
    if block == DEFAULT_DEPTH {
        return Some(format!("--depth-function {}", function.0));
    }
    Some(format!(
        "--depth-function {} --depth-block {}",
        function.0, block.0
    ))
}

#[cfg(test)]
#[path = "format_depth_query_test.rs"]
mod format_depth_query_test;
